Mathlib/Data/List/GroupBy.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2024 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.Data.List.Chain
+
+/-!
+# Group a list by a relation
+
+This file provides the basic API for `List.groupBy`. The main results are the following:
+
+- `List.join_groupBy`: the lists in `List.groupBy` join to the original list.
+- `List.nil_not_mem_groupBy`: the empty list is not contained in `List.groupBy`.
+- `List.chain'_of_mem_groupBy`: any two adjacent elements in a list in `List.groupBy` are related by
+  the specified relation.
+- `List.chain'_last_ne_head_groupBy`: the last element of each list in `List.groupBy` is not related
+  to the first element of the next list.
+
+Note that `List.groupBy` is defined in terms of a bespoke `List.groupBy.loop`. The arguments of
+`List.groupBy.loop l a g gs` represent the following:
+
+- `l : List α` is the list whose elements we group.
+- `a : α` is the previous element for which a comparison was performed.
+- `g : List α` is the group currently being assembled, in **inverse order**.
+- `gs : List (List α)` is all of the groups that have been completed, **in inverse order**.
+-/
+
+namespace List
+
+variable {α : Type*} {m : List α}
+
+@[simp]
+theorem groupBy_nil (r : α → α → Bool) : groupBy r [] = [] :=
+  rfl
+
+private theorem join_groupBy_loop (r : α → α → Bool) (l : List α) (a : α) (g : List α)
+    (gs : List (List α)) :
+    (groupBy.loop r l a g gs).join = gs.reverse.join ++ g.reverse ++ a::l := by
+  revert a g gs
+  induction l with
```

Comment:
Use `generalizing` instead of the revert/intro dance.

Reply:
Nice! I hadn't quite grasped what that did until now.